internal/app/machined/pkg/controllers/cri/image_cache_config.go

```diff
@@ -326,3 +440,105 @@ func (ctrl *ImageCacheConfigController) getImageCacheRoot(ctx context.Context, r
 
 	return optional.Some(targetPath), true, nil
 }
+
+func (ctrl *ImageCacheConfigController) copyImageCache(ctx context.Context, logger *zap.Logger, source, target string) error {
+	logger.Info("copying image cache", zap.String("source", source), zap.String("target", target))
+
+	if ctrl.DisableCacheCopy {
+		// used for testing
+		return nil
+	}
+
+	var bytesCopied int64
+
+	if err := filepath.WalkDir(source, func(path string, d fs.DirEntry, err error) error {
```

Comment:
I guess you could use `os.CopyFS` here and wrap `fs.FS.Open` interface method to record `bytesCopied`.

Reply:
see the docs: https://pkg.go.dev/os#CopyFS

it doesn't do what we need in case of overwrites/same files